### PR TITLE
expression, plan: rewrite builtin function: IS TRUE && IS FALSE

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

File: expression/builtin_op_test.go [hunk 1]
```diff
@@ -476,3 +477,77 @@ func (s *testEvaluatorSuite) TestUnaryNot(c *C) {
...
+
+	for _, tc := range testCases {
+		isTrueSig, err := funcs[ast.IsTruth].getFunction(datumsToConstants(types.MakeDatums(tc.args...)), s.ctx)
+		c.Assert(err, IsNil)
+		c.Assert(isTrueSig, NotNil)
+		c.Assert(isTrueSig.isDeterministic(), IsTrue)
+
+		isTrue, err := isTrueSig.eval(nil)
+		c.Assert(err, IsNil)
+		c.Assert(isTrue, testutil.DatumEquals, types.NewDatum(tc.isTrue))
```
Comment (XuHuaiyu) [inline expression/builtin_op_test.go]: s/ NewDatum/ NewIntDatum(int64(tc.isTrue))

File: expression/builtin_op_test.go [hunk 2]
```diff
@@ -476,3 +477,77 @@ func (s *testEvaluatorSuite) TestUnaryNot(c *C) {
...
+
+	for _, tc := range testCases {
+		isFalseSig, err := funcs[ast.IsFalsity].getFunction(datumsToConstants(types.MakeDatums(tc.args...)), s.ctx)
+		c.Assert(err, IsNil)
+		c.Assert(isFalseSig, NotNil)
+		c.Assert(isFalseSig.isDeterministic(), IsTrue)
+
+		isFalse, err := isFalseSig.eval(nil)
+		c.Assert(err, IsNil)
+		c.Assert(isFalse, testutil.DatumEquals, types.NewDatum(tc.isFalse))
```
Comment (XuHuaiyu) [inline expression/builtin_op_test.go]: ditto

File: expression/builtin_op_test.go [hunk 3]
```diff
@@ -476,3 +477,77 @@ func (s *testEvaluatorSuite) TestUnaryNot(c *C) {
...
+	sc.IgnoreTruncate = true
+
+	testCases := []struct {
+		args    []interface{}
+		isTrue  interface{}
+		isFalse interface{}
+	}{
+		{
+			args:    []interface{}{int64(-12)},
+			isTrue:  int64(1),
```
Comment (XuHuaiyu) [inline expression/builtin_op_test.go]: remove int64()

**Comment by zz-jason** [comment]
@XuHuaiyu updated, PTAL

**Comment by zz-jason** [comment]
@jackysp @breeswish @winkyao PTAL

File: expression/builtin_op.go [hunk 4]
```diff
@@ -341,40 +346,118 @@ func (b *builtinRightShiftSig) evalInt(row []types.Datum) (int64, bool, error) {
...
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, argTp)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = 1
+
+	var sig builtinFunc
+	switch {
+	case c.op == opcode.IsTruth && argTp == tpReal:
```
Comment (XuHuaiyu) [inline expression/builtin_op.go]: two switch may be more explict?
siwtch c.op{
switch argtP{
}
}

Comment (zz-jason) [inline expression/builtin_op.go]: no need, it makes code complex

**Pull request merged**